File: netlify/functions/subscribe.py
import contextlib
import json
import os
import secrets
import smtplib
import urllib.parse
from email.message import EmailMessage

from _ayb import ayb_query, ayb_rows, sql_literal

FROM_EMAIL = os.environ.get("FROM_EMAIL", "")
REPLY_TO = os.environ.get("REPLY_TO", "")
BLOG_NAME = "N=1 (marcua's blog)"
BLOG_URL = "https://blog.marcua.net"

CORS_HEADERS = {
    "Access-Control-Allow-Origin": "*",
    "Access-Control-Allow-Headers": "Content-Type",
    "Access-Control-Allow-Methods": "POST, OPTIONS",
}


def _send_confirmation(email, unsub_url):
    host = os.environ["SMTP_HOST"]
    port = int(os.environ.get("SMTP_PORT", "587"))
    username = os.environ["SMTP_USERNAME"]
    password = os.environ["SMTP_PASSWORD"]

    msg = EmailMessage()
    msg["From"] = FROM_EMAIL
    msg["To"] = email
    msg["Subject"] = f"You're subscribed to {BLOG_NAME}"
    if REPLY_TO:
        msg["Reply-To"] = REPLY_TO
    msg["List-Unsubscribe"] = f"<{unsub_url}>"
    msg["List-Unsubscribe-Post"] = "List-Unsubscribe=One-Click"
    msg.set_content(
        f"Thank you for subscribing! You'll get an email whenever "
        f"I publish a new post.\n\n"
        f"Visit the blog: {BLOG_URL}\n\n"
        f"To unsubscribe: {unsub_url}"
    )

    if port == 465:
        smtp = smtplib.SMTP_SSL(host, port)
    else:
        smtp = smtplib.SMTP(host, port)
        smtp.starttls()
    try:
        smtp.login(username, password)
        smtp.send_message(msg)
    finally:
        with contextlib.suppress(Exception):
            smtp.quit()


def handler(event, context):
    if event.get("httpMethod") == "OPTIONS":
        return {"statusCode": 200, "headers": CORS_HEADERS, "body": ""}

    if event.get("httpMethod") != "POST":
        return {
            "statusCode": 405,
            "headers": CORS_HEADERS,
            "body": json.dumps({"error": "Method not allowed"}),
        }

    try:
        body = json.loads(event.get("body", "{}"))
    except (json.JSONDecodeError, TypeError):
        return {
            "statusCode": 400,
            "headers": CORS_HEADERS,
            "body": json.dumps({"error": "Invalid JSON"}),
        }

    email = (body.get("email") or "").strip()
    if not email or "@" not in email:
        return {
            "statusCode": 400,
            "headers": CORS_HEADERS,
            "body": json.dumps({"error": "A valid email is required"}),
        }

    try:
        token = secrets.token_urlsafe(32)

        # Re-subscribe if previously unsubscribed, otherwise insert new.
        existing = ayb_rows(
            f"SELECT id, unsubscribed_at FROM subscribers "
            f"WHERE email = {sql_literal(email)}"
        )
        if existing:
            ayb_query(
                f"UPDATE subscribers SET unsubscribed_at = NULL, "
                f"unsubscribe_token = {sql_literal(token)} "
                f"WHERE id = {int(existing[0]['id'])}"
            )
        else:
            ayb_query(
                f"INSERT INTO subscribers (email, unsubscribe_token) "
                f"VALUES ({sql_literal(email)}, {sql_literal(token)})"
            )

        unsub_url = (
            f"{BLOG_URL}/.netlify/functions/unsubscribe"
            f"?token={urllib.parse.quote(token)}"
        )
        _send_confirmation(email, unsub_url)

        return {
            "statusCode": 200,
            "headers": CORS_HEADERS,
            "body": json.dumps({"ok": True}),
        }
    except Exception as e:
        return {
            "statusCode": 500,
            "headers": CORS_HEADERS,
            "body": json.dumps({"error": str(e)}),
        }

File: netlify/functions/subscribe.js
const { AybClient } = require("@aybdb/client");
const crypto = require("crypto");
const { BLOG_NAME, getAybClient, getTransporter } = require("./_shared");

const EMAIL_RE = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;

exports.handler = async (event) => {
  const corsHeaders = {
    "Access-Control-Allow-Origin": "*",
    "Access-Control-Allow-Headers": "Content-Type",
    "Access-Control-Allow-Methods": "POST, OPTIONS",
  };

  if (event.httpMethod === "OPTIONS") {
    return { statusCode: 200, headers: corsHeaders, body: "" };
  }

  if (event.httpMethod !== "POST") {
    return {
      statusCode: 405,
      headers: corsHeaders,
      body: JSON.stringify({ error: "Method not allowed" }),
    };
  }

  let body;
  try {
    body = JSON.parse(event.body || "{}");
  } catch {
    return {
      statusCode: 400,
      headers: corsHeaders,
      body: JSON.stringify({ error: "Invalid JSON" }),
    };
  }

  const email = (body.email || "").trim();
  if (!email || !EMAIL_RE.test(email)) {
    return {
      statusCode: 400,
      headers: corsHeaders,
      body: JSON.stringify({ error: "A valid email is required" }),
    };
  }

  try {
    const db = getAybClient();
    const token = crypto.randomBytes(32).toString("base64url");
    const escaped = AybClient.escapeSQL(email);
    const tokenEscaped = AybClient.escapeSQL(token);

    const existing = await db.queryObjects(
      `SELECT id, confirmed_at, unsubscribed_at FROM subscribers WHERE email = '${escaped}'`
    );

    if (existing.length > 0) {
      const sub = existing[0];
      if (sub.confirmed_at && !sub.unsubscribed_at) {
        return {
          statusCode: 200,
          headers: corsHeaders,
          body: JSON.stringify({ ok: true, already: true }),
        };
      }
      await db.query(
        `UPDATE subscribers SET unsubscribed_at = NULL, confirmed_at = NULL, ` +
        `secret_token = '${tokenEscaped}' ` +
        `WHERE id = ${parseInt(sub.id, 10)}`
      );
    } else {
      await db.query(
        `INSERT INTO subscribers (email, secret_token) ` +
        `VALUES ('${escaped}', '${tokenEscaped}')`
      );
    }

    const blogUrl = process.env.BLOG_URL || "";
    const confirmUrl = `${blogUrl}/newsletter/confirm?token=${encodeURIComponent(token)}`;

    const transporter = getTransporter();
    await transporter.sendMail({
      from: process.env.FROM_EMAIL,
      replyTo: process.env.REPLY_TO || undefined,
      to: email,
      subject: `Confirm your subscription to ${BLOG_NAME}`,
      text: [
        `Please confirm your subscription to ${BLOG_NAME} by visiting this link:`,
        "",
        confirmUrl,
        "",
        "If you didn't request this, you can ignore this email.",
      ].join("\n"),
    });

    return {
      statusCode: 200,
      headers: corsHeaders,
      body: JSON.stringify({ ok: true }),
    };
  } catch {
    return {
      statusCode: 500,
      headers: corsHeaders,
      body: JSON.stringify({ error: "Something went wrong, please try again." }),
    };
  }
};
